oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/AuthenticatedPrincipalOAuth2AuthorizedClientRepositoryTests.java

```diff
@@ -0,0 +1,142 @@
+	public void loadAuthorizedClientWhenAuthenticatedPrincipalThenLoadFromService() {
+		this.setupAuthenticatedPrincipal();
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		this.authorizedClientRepository.loadAuthorizedClient(this.registrationId, authentication, this.request);
+		verify(this.authorizedClientService).loadAuthorizedClient(this.registrationId, this.principalName);
+	}
+
+	@Test
+	public void loadAuthorizedClientWhenAnonymousPrincipalThenLoadFromAnonymousRepository() {
+		this.setupAnonymousPrincipal();
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		this.authorizedClientRepository.loadAuthorizedClient(this.registrationId, authentication, this.request);
+		verify(this.anonymousAuthorizedClientRepository).loadAuthorizedClient(this.registrationId, authentication, this.request);
+	}
+
+	@Test
+	public void saveAuthorizedClientWhenAuthenticatedPrincipalThenSaveToService() {
+		this.setupAuthenticatedPrincipal();
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		OAuth2AuthorizedClient authorizedClient = mock(OAuth2AuthorizedClient.class);
+		this.authorizedClientRepository.saveAuthorizedClient(authorizedClient, authentication, this.request, this.response);
+		verify(this.authorizedClientService).saveAuthorizedClient(authorizedClient, authentication);
+	}
+
+	@Test
+	public void saveAuthorizedClientWhenAnonymousPrincipalThenSaveToAnonymousRepository() {
+		this.setupAnonymousPrincipal();
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		OAuth2AuthorizedClient authorizedClient = mock(OAuth2AuthorizedClient.class);
+		this.authorizedClientRepository.saveAuthorizedClient(authorizedClient, authentication, this.request, this.response);
+		verify(this.anonymousAuthorizedClientRepository).saveAuthorizedClient(authorizedClient, authentication, this.request, this.response);
+	}
+
+	@Test
+	public void removeAuthorizedClientWhenAuthenticatedPrincipalThenRemoveFromService() {
+		this.setupAuthenticatedPrincipal();
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		this.authorizedClientRepository.removeAuthorizedClient(this.registrationId, authentication, this.request, this.response);
+		verify(this.authorizedClientService).removeAuthorizedClient(this.registrationId, this.principalName);
+	}
+
+	@Test
+	public void removeAuthorizedClientWhenAnonymousPrincipalThenRemoveFromAnonymousRepository() {
+		this.setupAnonymousPrincipal();
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		this.authorizedClientRepository.removeAuthorizedClient(this.registrationId, authentication, this.request, this.response);
+		verify(this.anonymousAuthorizedClientRepository).removeAuthorizedClient(this.registrationId, authentication, this.request, this.response);
+	}
+
+	private void setupAuthenticatedPrincipal() {
```

Comment:
We should not set the SecurityContextHolder. Instead return the Authentication and use it directly.